Handle UPD 15,22,27 (#874)
digital-asset/docs.daml.com: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst

-  Upgrading interface and exception definitions

> -  Adding/removing an interface instance on a template

Reviewer: :+1:

been implemented, but may be implemented in future releases.

-  Retroactive interface instances are not compatible with SCU upgrades.

>    SCU partially covers this behaviour, allowing instances to be changed in an upgrade,

Author: Fair enough, I'll update that

Remove Retroactive Instances
----------------------------

> SCU is not compatible with retroactive instances, so if you are migrating from an LF1.15
> project that uses this, you'll need to move the instances to their respective templates

Reviewer: ```suggestion
SCU is not compatible with retroactive interface instances, so if you are migrating to SCU from an LF1.15
project that uses retroactive instances, you'll need to move the instances to their respective templates
```

> project that uses retroactive instances, you'll need to move the instances to their respective templates
> during the migration.

Reviewer 2: ```suggestion
project that uses retroactive instances, you must move the instances to their respective templates
during the migration.
```

Remove Retroactive Instances
----------------------------

> SCU is not compatible with retroactive instances, so these will need to be removed.

Reviewer 3: In which situation would we "remove" them rather than not introduce them? I guess if you're compiling some old code to LF 1.17?

Author: Yeah, this is more in the context of migration, but migration points to this list. I'll see if theres better wording

usually the complexities of live upgrades comes with your workflows, not the data
transformations themselves. You can use Daml Script (with Canton) to test some
isolated simple cases, but for thorough tests of you system using SCU, you should
prefer full `Workflow Testing <#workflow-testing>`__.

When testing your upgrades with Daml Script, we recommend you place your Daml Script tests
in a separate package which depends on all versions of your business logic. This testing
package should not be uploaded to the ledger if possible, as it depends on the ``daml-script-lts`` package.
This package emits a warning on the participant when uploaded, as it serves no purpose on a participant,

> cannot be fully removed (as with any package), and may not be uploadable to the ledger in future versions (Daml 3).

Reviewer: ```suggestion
cannot be fully removed (as with any package), and may not be uploadable to the ledger in future versions (Daml 3). More information about this limitation :ref:`here <upgrade_dont_upload_daml_script>`.
```

Reviewer: This reference will be there once https://github.com/digital-asset/docs.daml.com/pull/878 goes in

Author: This doesn't exist yet, the :ref: would error on build

Author: I suspect this PR will merge first, given it already has language team reviews. You may want to update this in your PR.
We'll see, race condition

Reviewer: Yeah I figured - left it here not to forget

Author: @dylant-da This PR has merged, you'll need to make this addition on your side

> When testing your upgrades with Daml Script, we recommend you place your Daml Script tests
> in a separate package which depends on all versions of your business logic. This testing
> package should not be uploaded to the ledger if possible, as it depends on the ``daml-script-lts`` package.
> This package emits a warning on the participant when uploaded, as it serves no purpose on a participant,
> cannot be fully removed (as with any package), and may not be uploadable to the ledger in future versions (Daml 3).

Reviewer 2: ```suggestion
We recommend placing your Daml Script tests
in a separate package which depends on all versions of your business logic when testing your upgrades with Daml Script. This testing
package should not be uploaded to the ledger if possible, as it depends on the ``daml-script-lts`` package.
This package emits a warning on the participant when uploaded, as it serves no purpose on a participant,
cannot be fully removed (as with any package), and may not be uploadable to the ledger in future versions (Daml 3).
```

> Depending on multiple versions of the same package will however face ambiguity issues with
> imports, you can resolve this using :ref:`module prefixes <module_prefixes>`:

Reviewer 2: ```suggestion
Depending on multiple versions of the same package does however face ambiguity issues with
imports. You can resolve these issues using :ref:`module prefixes <module_prefixes>`:
```

.. code:: yaml

  name: my-testing-package
  ...
  data-dependencies:
    - my/v1/main-package.dar
    - my/v2/main-package.dar
  module-prefixes:
    main-1.0.0.dar: V11
    main-1.1.0.dar: V12

> For writing your tests, it is important that you verify old workflows are still functional under
> new data, and new implementation. You'll also need to verify that new workflows (that are intended
> to be backwards compatible) are able to consume old data. You should build your testing structure to
> cover this how you see fit, but we give the following recommendation as a starting point:

Reviewer 2: ```suggestion
It is important to verify old workflows are still functional under
new data and new implementation when writing your tests. You also need to verify that any new workflows intended
to be backward compatible can consume old data. You should build your testing structure to
cover this how you see fit, but we give the following recommendation as a starting point:
```

> If your new version only includes choice body or interface instance changes (i.e. it is a patch release)

Reviewer 2: ```suggestion
If your new version only includes choice body or interface instance changes (that is, it is a patch release)
```

> -  | Simply run your existing test suite written for V1 but updated to call V2 choices. This can be

Reviewer 2: ```suggestion
-  | Run your existing test suite for V1 but updated to call V2 choices. This can be
```

     done with a rewrite, or by passing down a :ref:`package preference <daml-script-package-preference>`
     and calling the test with both the V1 and V2 package ID.

> If your new version includes data changes, be that to contract payloads or choices (i.e. it is a minor release)

Reviewer 2: ```suggestion
If your new version includes data changes, be that to contract payloads or choices (that is, it is a minor release)
```

usually the complexities of live upgrades comes with your workflows, not the data
transformations themselves. You can use Daml Script (with Canton) to test some
isolated simple cases, but for thorough tests of you system using SCU, you should
prefer full `Workflow Testing <#workflow-testing>`__.

When testing your upgrades with Daml Script, we recommend you place your Daml Script tests
in a separate package which depends on all versions of your business logic. This testing

> package should not be uploaded to the ledger if possible, as it cannot be dropped in an upgrade.

Reviewer 3: "as it cannot be dropped in an upgrade": technically you can no? Since it's a separate package that presumably no other package depends on, you can choose not to upload any upgrade to it. You can even unvet it.

It seems to me the real reason we don't want users to upload such a package is because it depends on daml-script and we don't want users to upload daml-script-lts.

Author: This is true, but given unvetting isn't deletion, they would still be stuck with it on their ledger.
I'll try a better wording, but i really do mean in general, dont upload daml script

Reviewer 3: Oh absolutely! My point was maybe what we should say just that: don't upload that package because it will upload daml-script and we don't want that.

If your new version only includes choice body or interface instance changes (i.e. it is a patch release)

-  | Simply run your existing test suite written for V1 but updated to call V2 choices. This can be
     done with a rewrite, or by passing down a :ref:`package preference <daml-script-package-preference>`
     and calling the test with both the V1 and V2 package ID.

If your new version includes data changes, be that to contract payloads or choices (i.e. it is a minor release)

-  | Assuming your data change affects a template payload, write separate setup code for V1 and V2, populating new fields
   | ``setupV1 : Script V1TestData``
   | ``setupV2 : Script V2TestData``
   | These new data types should mostly just hold Contract IDs

-  | Update your existing test suite from V1 to take a :ref:`package preference <daml-script-package-preference>`,
     allowing the V2 implementation without additional fields to choices to be called.
   | ``testV1 : [PackageId] -> V1TestData -> Script ()``

-  | Run the above test suite against V1 data, passing a V1 preference, then a V2 preference.
   | This ensures your changes haven't broken any existing workflows.

> -  | Next write tests for your new/modified workflows, using the V2 choice implementations. This doesn't not need a package preference.

Reviewer 3: ```suggestion
-  | Next write tests for your new/modified workflows, using the V2 choice implementations. This doesn't need a package preference.
```

>    | ``testV2 : TestData -> Script ()``

Reviewer 3: Should that be V2TestData?

-  | Run these tests against both the V1 setup and the V2 setup, to ensure your new workflows support existing/old templates.
   | In order to do this, you'll need some way to upcast your ``V1TestData``, i.e.
   | ``upcastTestData : V1TestData -> V2TestData``
   | This function should mostly just call ``coerceContractId`` on any contract IDs, and fill in any ``None`` values if needed.

> -  | Finally, you can cover any workflows that require the contract data to already be upgraded (new fields populated), these are

Reviewer 3: Not an english expert but is that correct? Shouldn't that either be "any workflow that requires" or "worflows that require"?

Author: I think this is fine, for example you can say like "Do you have any ideas". Its okay to follow `any` by a plural

Author: I could reword it to `all workflows`, I'll wait to see what the docs team thinks

Reviewer 3: If it sounds good to you, an english native speaker, then I think that's ok :)

Reviewer: I think it's OK, there's a small but in this case meaningless difference I think.

Reviewer 2: "any workflows" and "all workflows" are equally correct in this case, since "any" implies it could be one or many.
